File: Calculator/android-native/app/src/main/java/com/qaantum/calculatorhub/calculators/RetirementCalculator.kt
package com.qaantum.calculatorhub.calculators

import kotlin.math.pow

data class RetirementResult(
    val totalSavings: Double,
    val monthlyIncome: Double // Based on 4% rule
)

class RetirementCalculator {
    fun calculate(
        currentAge: Int,
        retirementAge: Int,
        currentSavings: Double,
        monthlyContribution: Double,
        annualRate: Double
    ): RetirementResult? {
        val years = retirementAge - currentAge
        val months = years * 12
        
        if (months <= 0) return null

        val rate = annualRate / 100.0 / 12.0

        // FV of Initial Savings
        val fvSavings = currentSavings * (1 + rate).pow(months.toDouble())

        // FV of Contributions
        val fvContributions = if (rate != 0.0) {
            monthlyContribution * ((1 + rate).pow(months.toDouble()) - 1) / rate
        } else {
            monthlyContribution * months
        }

        val total = fvSavings + fvContributions
        // 4% Rule for safe withdrawal rate (annual) -> monthly
        val monthlyIncome = (total * 0.04) / 12.0

        return RetirementResult(
            totalSavings = total,
            monthlyIncome = monthlyIncome
        )
    }
}

File: Calculator/ios-native/CalculatorHub/Calculators/RetirementCalculator.swift
import Foundation

struct RetirementResult {
    let totalSavings: Double
    let monthlyIncome: Double
}

class RetirementCalculator {
    func calculate(currentAge: Int, retirementAge: Int, currentSavings: Double, monthlyContribution: Double, annualRate: Double) -> RetirementResult? {
        let years = retirementAge - currentAge
        let months = years * 12
        
        guard months > 0 else { return nil }

        let rate = annualRate / 100.0 / 12.0
        let fvSavings = currentSavings * pow(1 + rate, Double(months))
        
        let fvContributions: Double
        if rate != 0 {
            fvContributions = monthlyContribution * (pow(1 + rate, Double(months)) - 1) / rate
        } else {
            fvContributions = monthlyContribution * Double(months)
        }

        let total = fvSavings + fvContributions
        let monthlyIncome = (total * 0.04) / 12.0

        return RetirementResult(totalSavings: total, monthlyIncome: monthlyIncome)
    }
}
